symbolic-debuginfo/src/sourcebundle/utf8_reader.rs

```diff
@@ -0,0 +1,282 @@
+//! UTF-8 reader used by the sourcebundle module to read files.
+
+use std::cmp;
+use std::io::{Error, ErrorKind, Read, Result};
+use std::str;
+
+use thiserror::Error;
+
+const MAX_UTF8_SEQUENCE_SIZE: usize = 4;
+
+#[derive(Debug, Error)]
+#[error("Invalid UTF-8 sequence")]
+pub(crate) struct UTF8ReaderError;
+
+pub struct Utf8Reader<R> {
+    inner: R,
+
+    /// A buffer of `MAX_UTF8_SEQUENCE_SIZE` bytes, which we use when the end of a read is not a
+    /// valid UTF-8 sequence. We read into this buffer until we have a valid UTF-8 sequence, or
+    /// the reader is exhausted, or the buffer is full. Assuming we get a valid UTF-8 sequence,
+    /// we then on next read read from this buffer.
+    buffer: [u8; MAX_UTF8_SEQUENCE_SIZE],
+
+    /// The index of the first byte in the buffer that has not been read.
+    buffer_start: usize,
+
+    /// The index of the last byte in the buffer that is part of the UTF-8 sequence. We only read
+    /// up to this index, bytes after this index are discarded.
+    buffer_end: usize,
+}
+
+impl<R> Utf8Reader<R> {
+    pub fn new(inner: R) -> Self {
+        Self {
+            inner,
+            buffer: [0; MAX_UTF8_SEQUENCE_SIZE],
+            buffer_start: 0,
+            buffer_end: 0,
+        }
+    }
+
+    /// The slice of the buffer that should be read next.
+    fn buffer_to_read(&self) -> &[u8] {
+        &self.buffer[self.buffer_start..self.buffer_end]
+    }
+
+    /// Advances the buffer start index by the given amount.
+    fn advance(&mut self, amt: usize) {
+        self.buffer_start += amt;
+    }
+
+    /// Reads bytes from the buffer into the given buffer.
+    fn read_from_buffer(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let bytes_copied = slice_copy(buf, self.buffer_to_read());
+        self.advance(bytes_copied);
+
+        Ok(bytes_copied)
+    }
+}
+
+impl<R> Utf8Reader<R>
+where
+    R: Read,
+{
+    /// Reads bytes from the inner reader into the given buffer, if needed, filling self.buffer
+    /// with the remaining bytes of the UTF-8 sequence at the end of the buffer, which did not fit
+    /// in the read.
```

Comment:
This docstring can be simplified a bit.
```suggestion
    /// Reads bytes from the inner reader into the given buffer.
    ///
    /// If a UTF-8 sequence only partially fits into the buffer, the remaining
    /// bytes are written to `self.buffer`.
```